### ISSUE TITLE
[Question] API GW V2 metrics not found with dimensionNameRequirements

### ISSUE THREAD

**Issue by bogdankatishev** [nerdswords/yet-another-cloudwatch-exporter#1031]
Hello,

I have a question regarding the scraping of API GW V2 metrics. I noticed that the supporting PR was released in v0.52.0 for scraping API GW  V2 metrics (HTTP API).

We have this config block in our config.yaml:
```
      # API GW V1 (REST)
      # https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-metrics-and-dimensions.html
      - <<: *default_job_config
        type: AWS/ApiGateway
        dimensionNameRequirements:
          - ApiName
          - Stage
        metrics:
          - name: Latency
            statistics: [ Average, Maximum, p95, p99 ]
          - name: Count
            statistics: [ SampleCount, Sum ]
          - name: 4XXError
            statistics: [ Sum, Average ]
          - name: 5XXError
            statistics: [ Sum, Average ]

      # API GW V2 (HTTP)
      # https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-metrics.html
      - <<: *default_job_config
        type: AWS/ApiGateway
        dimensionNameRequirements:
          - ApiId
          - Stage
          - Route
        metrics:
          - name: Latency
            statistics: [ Average, Maximum, p95, p99 ]
          - name: Count
            statistics: [ SampleCount, Sum ]
          - name: 4xx
            statistics: [ Sum ]
          - name: 5xx
            statistics: [ Sum ]
```

default_job_config has the values: 
```
default_job_config: &default_job_config
  addCloudwatchTimestamp: true
  roles:
    - <roles>
  regions:
    - eu-west-1
  period: 300
  length: 300
```

We are also using these feature flags: 
```
-enable-feature=aws-sdk-v2,max-dimensions-associator,list-metrics-callback
```

But we are getting these logs when running (multiple times):
```
{"account":"<account>","arn":"<role>","job_type":"AWS/ApiGateway","level":"info","msg":"No metrics data found","region":"eu-west-1","time":"2023-06-22T13:09:42Z"}
```

What are we doing wrong in our config.yaml?

**Comment by bogdankatishev** [comment]
Hello @ash-cohesity,

We also tried this without using `dimensionNameRequirements`. But we were still missing `4xx` and `5xx` metrics from API GW V2.

And we also do not want to use `searchTags` because we want to scrape all API GW's in that account.

**Comment by wimsymons** [comment]
#816 is a duplicate and there is a possible fix in #984

**Comment by cristiangreco** [comment]
> We also tried this without using `dimensionNameRequirements`. But we were still missing `4xx` and `5xx` metrics from API GW V2.

@bogdankatishev Are those metrics missing only from YACE, e.g. do you see those metrics in CloudWatch console? Note that some combination of dimensions require an additional setting to be enabled.

<img width="1160" alt="Screenshot 2023-06-27 at 11 13 32" src="https://github.com/nerdswords/yet-another-cloudwatch-exporter/assets/316923/5dcfb1bc-ea7a-4163-adcd-8a94a0379198">

**Comment by bogdankatishev** [comment]
@cristiangreco Yes these metrics are only missing from YACE, we see them in the CloudWatch console. 
We are also aware that some combinations require an additional setting to be enabled. We have this setting enabled on the API GW.

**Comment by flomsk** [comment]
Also try to setup ApiGW metrics with `dimensionNameRequirements` , and also receive `msg":"skipping metric unmatched by associator`

No feature flag helped, neither `aws-sdk-v2` nor `max-dimensions-associator` nor both.

**Comment by cristiangreco** [comment]
@flomsk a debug log would help here, thanks!

**Comment by cristiangreco** [comment]
Thanks @flomsk, do you also have the part of the output relative to the GetTagging API request? Also, can you please provide examples of ARN names for your apigw resources? Thanks so much!

**Comment by cristiangreco** [comment]
@flomsk just by looking at the logs you pasted above, I would expect that by removing  `dimensionNameRequirements` you would still get the `4xx`  and `5xx` metrics with `ApiId` and `Stage` dimensions.

**Comment by flomsk** [comment]
hey @cristiangreco 
im not sure i took the correct one about GetTagging, is that what you need?
```
-----------------------------------------------------
2023/07/06 09:08:22 {"PaginationToken":"","ResourceTagMappingList":[{"ResourceARN":"arn:aws:apigateway:eu-west-1::/apis/xxxxxxxxxxx/stages/beta","Tags":[{"Key":"app","Value":"testapp"}]},{"ResourceARN":"arn:aws:apigateway:eu-west-1::/apis/xxxxxxxxxxx","Tags":[{"Key":"app","Value":"testapp"}]}]}
{"level":"debug","msg":"GetResourcesPages finished","time":"2023-07-06T09:08:22Z","total":2}
```

also removing `dimensionNameRequirements` giving me only one dimension `ApiId`
```
aws_apigateway_5xx_sum{account_id="123456789012",dimension_ApiId="xxxxxxxxxxx",name="arn:aws:apigateway:eu-west-1::/apis/xxxxxxxxxxx",region="eu-west-1"}
```

And also you can see ARN on the lines above

**Comment by cristiangreco** [comment]
I think this is kinda expected behaviour for the way the associator algorithm work -- basically it's not able to associate the metric with multiple dimensions (`ApiId`, `Stage`, `Resource`, `Method`) back to the ARN of the gw (which contains only the `ApiId` value). That's why you see only `ApiId` in the `4xx` and `5xx` metrics.

I don't think `dimensionNameRequirements` should be used here, it's pointless as it only filters metrics _after_ the associator.

I can try to think of ways to change the algorithm to account for cases like this, but don't have a way to easily test it at the moment as I don't have any apigateway resource.

**Comment by cristiangreco** [comment]
@flomsk are you able to test a custom branch or build? Would you please be able to let me know what happens with this PR? https://github.com/nerdswords/yet-another-cloudwatch-exporter/pull/1067

Please continue to use the `max-dimensions-associator` feature flag.

**Comment by bogdankatishev** [comment]
Hello @cristiangreco 

We have tested it again with the 0.53.0 release with the feature flag: `-enable-feature=max-dimensions-associator` but still no luck. These dimensions are missing as labels in the prometheus metrics: method, stage and resource when using the discovery job config.

When using customNamespace method to fetch ApiGateway metrics, we receive all the needed labels correctly (method, stage and resource):
```
customNamespace:
- addCloudwatchTimestamp: true
  regions:
  - eu-west-1
  period: 300
  length: 300
  name: temporary-apigateway
  namespace: AWS/ApiGateway
  metrics:
  - name: Latency
    statistics:
    - Average
    - Maximum
    - p95
    - p99
  - name: Count
    statistics:
    - SampleCount
    - Sum
  - name: 4xx
    statistics:
    - Sum
  - name: 4XXError
    statistics:
    - Sum
    - Average
  - name: 5xx
    statistics:
    - Sum
  - name: 5XXError
    statistics:
    - Sum
    - Average

```
```
aws_apigateway_4xx_sum{account_id="xxx", dimension_api_id="xxx", dimension_method="GET", dimension_resource="/path/of/resource", dimension_stage="$default", env="xxx", job="xxx", name="temporary-apigateway", region="eu-west-1"}
```

BUT, when using the discovery method to fetch ApiGateway metrics, we lose all those labels (method, stage and resource):
```
---
apiVersion: v1alpha1
discovery:
  jobs:
  - addCloudwatchTimestamp: true                  
    regions:                                      
    - eu-west-1                                   
    period: 300                                   
    length: 300                                   
    type: AWS/ApiGateway                          
    metrics:                                      
    - name: Latency                               
      statistics:                                 
      - Average                                   
      - Maximum                                   
      - p95                                       
      - p99                                       
    - name: Count                                 
      statistics:                                 
      - SampleCount                               
      - Sum                                       
    - name: 4xx                                   
      statistics:                                 
      - Sum                                       
    - name: 4XXError                              
      statistics:                                 
      - Sum                                       
      - Average                                   
    - name: 5xx                                   
      statistics:                                 
      - Sum                                       
    - name: 5XXError                              
      statistics:                                 
      - Sum                                       
      - Average  
```
```
aws_apigateway_4xx_sum{account_id="xxx", dimension_api_id="xxx", env="xxx", job="xxx", name="arn:aws:apigateway:eu-west-1::/apis/xxx", region="eu-west-1"}
```

**Note that we are not using: `dimensionNameRequirements` anymore :arrow_double_up:** 

We also did not test #1067 yet.

Debug output is also the same as @flomsk has pasted above already.

**Comment by bogdankatishev** [comment]
Hello @cristiangreco, 

I just tested #1067 and it seems to work! In the debug output, the labels: Method, Stage and Resource are being registered and shown correctly! :)

**Comment by cristiangreco** [comment]
@bogdankatishev did you use `max-dimensions-associator` when testing https://github.com/nerdswords/yet-another-cloudwatch-exporter/pull/1067?

**Comment by bogdankatishev** [comment]
@cristiangreco Yes, i used: `./yace -debug -enable-feature=max-dimensions-associator`

**Comment by cristiangreco** [comment]
@bogdankatishev thanks for the confirmation -- that's a bit strange because it works ok for me locally. 

Do you see the `yace_cloudwatch_apigatewayapi_requests_total` counter in the `/metrics` page?
Would you be able to paste more debug logs perhaps?

**Comment by cristiangreco** [comment]
@bogdankatishev can you please share the full content (all labels) of the `aws_apigateway_info` metric?

**Comment by cristiangreco** [comment]
@bogdankatishev is your gw using only the `$default` stage? Could you make sure that the stage is tagged as well?

You should get something similar (in my case I'm using `gw_has_tag=yes` on the gateway and `stage_has_tag=yes` on the stage):

```
# HELP aws_apigateway_4xx_sum Help is not implemented yet.
# TYPE aws_apigateway_4xx_sum gauge
aws_apigateway_4xx_sum{account_id="1234567890",dimension_ApiId="abc123",dimension_Stage="",name="arn:aws:apigateway:eu-west-3::/apis/abc123",region="eu-west-3"} NaN
aws_apigateway_4xx_sum{account_id="1234567890",dimension_ApiId="abc123",dimension_Stage="$default",name="arn:aws:apigateway:eu-west-3::/apis/abc123/stages/$default",region="eu-west-3"} NaN
# HELP aws_apigateway_info Help is not implemented yet.
# TYPE aws_apigateway_info gauge
aws_apigateway_info{name="arn:aws:apigateway:eu-west-3::/apis/abc123",tag_gw_has_tag="yes",tag_stage_has_tag=""} 0
aws_apigateway_info{name="arn:aws:apigateway:eu-west-3::/apis/abc123/stages/$default",tag_gw_has_tag="",tag_stage_has_tag="yes"} 0
```

**Comment by bogdankatishev** [comment]
> @bogdankatishev can you please share the full content (all labels) of the `aws_apigateway_info` metric?

Yes:
```
aws_apigateway_info{name="arn:aws:apigateway:eu-west-1::/apis/xxx",tag_SolutionId="",tag_afdeling="xxx",tag_availability="",tag_aws_cloudformation_logical_id="",tag_aws_cloudformation_stack_id="",tag_aws_cloudformation_stack_name="",tag_httpapi_createdBy="",tag_kp="xxx",tag_owner="xxx",tag_project="xxx"}
```



> @bogdankatishev is your gw using only the `$default` stage? Could you make sure that the stage is tagged as well?
> 
> You should get something similar (in my case I'm using `gw_has_tag=yes` on the gateway and `stage_has_tag=yes` on the stage):
> 
> ```
> # HELP aws_apigateway_4xx_sum Help is not implemented yet.
> # TYPE aws_apigateway_4xx_sum gauge
> aws_apigateway_4xx_sum{account_id="1234567890",dimension_ApiId="abc123",dimension_Stage="",name="arn:aws:apigateway:eu-west-3::/apis/abc123",region="eu-west-3"} NaN
> aws_apigateway_4xx_sum{account_id="1234567890",dimension_ApiId="abc123",dimension_Stage="$default",name="arn:aws:apigateway:eu-west-3::/apis/abc123/stages/$default",region="eu-west-3"} NaN
> # HELP aws_apigateway_info Help is not implemented yet.
> # TYPE aws_apigateway_info gauge
> aws_apigateway_info{name="arn:aws:apigateway:eu-west-3::/apis/abc123",tag_gw_has_tag="yes",tag_stage_has_tag=""} 0
> aws_apigateway_info{name="arn:aws:apigateway:eu-west-3::/apis/abc123/stages/$default",tag_gw_has_tag="",tag_stage_has_tag="yes"} 0
> ```

Yes the gw only has the `$default` stage, but I see that our stage is indeed not tagged. 
Seems like the tagging propagation of AWS only goes to the API GW resource and not further to the stages.

Okay but even if we start tagging the stage, if I understand correctly, we are still missing the `method` and `resource` label from the GW metrics when using the `max-dimensions-associator` feature flag.

**We also found out that when **NOT** using the `max-dimensions-associator` feature flag, we get all the labels correctly (method/stage/resource) on the `/metrics` page (even if the stage is not tagged).**
`./yace -debug -enable-feature=aws-sdk-v2`
OR 
`./yace -debug`
```
aws_apigateway_4xx_sum{account_id="xxx",dimension_ApiId="xxx",dimension_Method="",dimension_Resource="",dimension_Stage="",name="arn:aws:apigateway:eu-west-1::/apis/xxx",region="eu-west-1"} 0 1690451100000
aws_apigateway_4xx_sum{account_id="xxx",dimension_ApiId="xxx",dimension_Method="",dimension_Resource="",dimension_Stage="$default",name="arn:aws:apigateway:eu-west-1::/apis/xxx",region="eu-west-1"} 0 1690451100000
aws_apigateway_4xx_sum{account_id="xxx",dimension_ApiId="xxx",dimension_Method="GET",dimension_Resource="/public/{proxy+}",dimension_Stage="$default",name="arn:aws:apigateway:eu-west-1::/apis/xxx",region="eu-west-1"} 0 1690451100000
aws_apigateway_4xx_sum{account_id="xxx",dimension_ApiId="xxx",dimension_Method="OPTIONS",dimension_Resource="/{proxy+}",dimension_Stage="$default",name="arn:aws:apigateway:eu-west-1::/apis/xxx",region="eu-west-1"} 0 1690451100000
```

So to recap @cristiangreco : when **NOT** using `max-dimensions-associator` feature flag, all of the API GW metrics and labels/dimensions are being captured correctly :partying_face: (tested on: #1067)

**Comment by bogdankatishev** [comment]
@cristiangreco Could you also please explain what this code does? https://github.com/nerdswords/yet-another-cloudwatch-exporter/blob/cristian/rm-apigw-filter/pkg/clients/tagging/v2/filters.go#L31-L82

What does this filter do/Why do we need filtering?

**Comment by cristiangreco** [comment]
@bogdankatishev thanks for the feedback!

> Okay but even if we start tagging the stage, if I understand correctly, we are still missing the method and resource label from the GW metrics when using the max-dimensions-associator feature flag.

That's a bit strange, I don't see this difference between the base and max-dim associator.

>We also found out that when NOT using the max-dimensions-associator feature flag, we get all the labels correctly (method/stage/resource) on the /metrics page (even if the stage is not tagged).

Was this happening before I suggested to add a tag to the stage as well? Or did it start happening after you removed the tag? The best test would be to create a new ApiGatewa, leave the `$default` stage without tags and test yace both with and withouth the feature flag.

**Comment by bogdankatishev** [comment]
> Was this happening before I suggested to add a tag to the stage as well? Or did it start happening after you removed the tag? The best test would be to create a new ApiGatewa, leave the $default stage without tags and test yace both with and withouth the feature flag.

@cristiangreco We never added tags to our stage, so we tested on a clean API GW with a $default stage that never had tags.
Me and another co-worker built #1067 and tested it with and without the feature flag and we got the same results: when not using the max-dimensions-associator feature flag, we got correct results on the `/metrics` page.

> That's a bit strange, I don't see this difference between the base and max-dim associator.

Are you sure that you are on #1067 build? Because me and another co-worker are able to reproduce these findings ⬆️

**Comment by cristiangreco** [comment]
I've dumped a potential fix for the feature flag `maxdimassociator` in https://github.com/nerdswords/yet-another-cloudwatch-exporter/pull/1081

It seems to work correctly now, but you need https://github.com/nerdswords/yet-another-cloudwatch-exporter/pull/1067 as well. I'll try to get both merged tomorrow and so you can test a build from `master`.

**Comment by benbridts** [comment]
I've rebased #1067 on master and can confirm that the dimensions are present with and without `--enable-feature=aws-sdk-v2,max-dimensions-associator,list-metrics-callback`.

eg:

```
aws_apigateway_4xx_sum{account_id="XXX",dimension_ApiId="YYY",dimension_Method="",dimension_Resource="",dimension_Stage="",name="arn:aws:apigateway:eu-west-1::/apis/YYY",region="eu-west-1"} 452 1692179100000
aws_apigateway_4xx_sum{account_id="XXX",dimension_ApiId="YYY",dimension_Method="",dimension_Resource="",dimension_Stage="$default",name="arn:aws:apigateway:eu-west-1::/apis/YYY",region="eu-west-1"} 413 1692178800000
aws_apigateway_4xx_sum{account_id="XXX",dimension_ApiId="XXX",dimension_Method="GET",dimension_Resource="/ZZZ/graphql/public/v1",dimension_Stage="$default",name="arn:aws:apigateway:eu-west-1::/apis/YYY",region="eu-west-1"} 0 -62135596800000
```

Doing the same thing on  the master branch, also gives me the dimensions if I keep the feature flag enabled, but not without it. It also returns a bunch more metrics.

Putting that in in table:

branch | flags | aws_apigateway metrics returned | dimension_ApiId | dimension_Method
---|---|---|---|---
master | no | 1588 | ✅  | ❌ 
master | yes | 1692 | ✅ | ✅ 
(rebased) rm-apigw-filter | no | 325 | ✅ | ✅ 
(rebased) rm-apigw-filter | yes | 309 | ✅ | ✅ 


I took a quick look  to try explain that difference in metrics, and looked at the different kind of metrics but it's hard for me to know what the expected numbers are (because I assume some will have older data points, so I can't go by the number of APIs that we have).

It is interesting to see that with my configuration, the `aws_apigateway_4_xxerror_average` does not exist anymore in the new branch

metric | master/no-flags | master/flags | rm-apigw-filter/no-flags | rm-apigw-filter/flags
---|---|---|---|---
aws_apigateway_4xx_sum | 8 | 21 | 23 | 21
aws_apigateway_5xx_sum | 8 | 21 | 23 | 21
aws_apigateway_latency_average |  155 | 168 | 23 | 21
aws_apigateway_latency_maximum | 155 | 168 | 23 | 21
aws_apigateway_latency_p95 | 155 | 168 | 23 | 21
aws_apigateway_latency_p99 | 155 | 168 | 23 | 21
aws_apigateway_count_sample_count | 156 | 169 | 24 | 22
aws_apigateway_count_sum |  156 | 169 | 24 | 22
aws_apigateway_info  | 52 | 52| 139 | 139
aws_apigateway_4_xxerror_average |  147 | 147 | - | -
aws_apigateway_4_xxerror_sum |  147 | 147 | - | -
aws_apigateway_5_xxerror_average | 147 | 147 | - | -
aws_apigateway_5_xxerror_sum | 147 | 147 | - | -

**Comment by bogdankatishev** [comment]
Any news on this? @cristiangreco
